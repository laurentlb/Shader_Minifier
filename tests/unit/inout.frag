// Multifile test, part 1

// The test is to ensure that exported variables are renamed in the same way
// across multiple files.

#version 330

// These 4 variables should have the same name as in the other file.
uniform samplerCube texture0;

in vec3 normal;
in vec3 viewVec;

out vec4 fragmentColor;

void main()
{
  vec3 E = normalize(viewVec);
  vec3 N = normalize(normal);
  vec3 albedo = vec3(0.1, 0.2, 0.3);
  vec3 f0 = vec3(0.5, 0.5, 0.5);

  float IoR = 1.5;
  vec3 reflection = texture(texture0, reflect(-E, N)).rgb;
  vec3 refraction = texture(texture0, refract(-E, N, 1./IoR)).rgb;

  vec3 color = mix(albedo * refraction, reflection, 0.1);
  fragmentColor = vec4(color, 1.);
}

// These functions are same as in the other file. Ideally (for better compression),
// the local variables and arguments should remain the same.
vec3 Schlick(vec3 f0, vec3 E, vec3 H)
{
  float x = 1. - clamp(dot(E, H), 0., 1.);
  return x*x*x*x*x * (1. - f0) + f0;
}

vec3 BlinnPhong(vec3 E, vec3 L, vec3 N, vec3 albedo, vec3 f0, float roughness)
{
  vec3 H = normalize(E + L);
  float alpha = 1. + 2048. * (1. - roughness)*(1. - roughness);

  vec3 diffuse = albedo;
  vec3 specular = vec3(pow(clamp(dot(H, N), 0., 1.), alpha) * (alpha + 4.) / 8.);
  vec3 fresnel = Schlick(f0, E, H);

  return mix(diffuse, specular, fresnel);
}
