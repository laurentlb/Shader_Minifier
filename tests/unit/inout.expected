// Generated with  (https://github.com/laurentlb/Shader_Minifier/)
#ifndef SHADER_MINIFIER_IMPL
#ifndef SHADER_MINIFIER_HEADER
# define SHADER_MINIFIER_HEADER
# define VAR_ambientLight "r"
# define VAR_diffuseColor "c"
# define VAR_emissiveColor "p"
# define VAR_fragmentColor "f"
# define VAR_mediumDensity "s"
# define VAR_normal "m"
# define VAR_specularColor "i"
# define VAR_texture0 "v"
# define VAR_viewVec "d"
#endif

#else // if SHADER_MINIFIER_IMPL

// tests/unit/inout.frag
"#version 330\n"
 "uniform samplerCube v;"
 "in vec3 m,d;"
 "out vec4 f;"
 "void main()"
 "{"
   "vec3 c=normalize(d),r=normalize(m);"
   "f=vec4(mix(vec3(.1,.2,.3)*texture(v,refract(-c,r,1./1.5)).xyz,texture(v,reflect(-c,r)).xyz,.1),1);"
 "}"
 "vec3 e(vec3 v,vec3 m,vec3 r)"
 "{"
   "float f=1.-clamp(dot(m,r),0.,1.);"
   "return f*f*f*f*f*(1.-v)+v;"
 "}"
 "vec3 e(vec3 v,vec3 m,vec3 r,vec3 c,vec3 d,float f)"
 "{"
   "vec3 p=normalize(v+m),i;"
   "float s=1.+2048.*(1.-f)*(1.-f);"
   "i=e(d,v,p);"
   "return mix(c,vec3(pow(clamp(dot(p,r),0.,1.),s)*(s+4.)/8.),i);"
 "}",

// tests/unit/inout2.frag
"#version 330\n"
 "uniform samplerCube v;"
 "uniform float s;"
 "uniform vec3 r,c,p,i;"
 "in vec3 m,d;"
 "out vec4 f;"
 "vec3 e(vec3 v,vec3 m,vec3 r)"
 "{"
   "float f=1.-clamp(dot(m,r),0.,1.);"
   "return f*f*f*f*f*(1.-v)+v;"
 "}"
 "void main()"
 "{"
   "f=vec4(p+mix(c*r,r,.5),1);"
 "}"
 "vec3 e(vec3 v,vec3 m,vec3 r,vec3 c,vec3 d,float f)"
 "{"
   "vec3 p=normalize(v+m),i;"
   "float s=1.+2048.*(1.-f)*(1.-f);"
   "i=e(d,v,p);"
   "return mix(c,vec3(pow(clamp(dot(p,r),0.,1.),s)*(s+4.)/8.),i);"
 "}",

#endif
